### PR TITLE
feat: UIPointerEvents implementation

### PR DESCRIPTION
**Pull request by sandrade-dcl**
### What this PR includes?
- Implemention of `onMouseDown` and `onMouseUp` events for SDK UI component.

### New systems implemented
- `UIPointerEventsSystem`

### CODE REVIEW

File: Explorer/Assets/DCL/SDKComponents/SceneUI/Systems/UIPointerEvents/UIPointerEventsSystem.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        {
+            if (!sdkModel.IsDirty)
+                return;
+
+            foreach (var pEvent in sdkModel.PointerEvents)
+            {
+                switch (pEvent.EventType)
+                {
+                    case PointerEventType.PetDown:
+                        uiTransformComponent.Transform.RegisterPointerDownCallback(_ => AppendMessage(sdkEntity, pEvent.EventInfo.Button, PointerEventType.PetDown));
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/SDKComponents/SceneUI/Systems/UIPointerEvents/UIPointerEventsSystem.cs:49]: You introduced a dangerous closure here: you capture `CRDTEntity` and `pEvent`. It results in:
- unnecessary allocations 
- disconnects from the component/entity life-cycle.

Event handling contradicts ECS lifecycle as it disconnects the moment of the callback from the system update order and its possible destruction.

It should be improved:
- I propose you to gather indices and `PointerEventType` being clicked in a separate structure (you may keep it in `UITransformComponent`), 
- Then in a regular query iterate over that data and call `AppendMessage` accordingly
- Thus you will be able to take all the required data from the query without closures

**Review by mikhail-dcl — CHANGES_REQUESTED** [review]

File: Explorer/Assets/DCL/SDKComponents/SceneUI/Classes/DCLTransform.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
+﻿using Arch.Core;
+using System.Collections.Generic;
+using UnityEngine.UIElements;
+
+namespace DCL.SDKComponents.SceneUI.Classes
+{
+    public class DCLTransform
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/SDKComponents/SceneUI/Classes/DCLTransform.cs]: Effectively, you were right! I've applied the refactor in order to make `UITransformComponent` a poolable class and everything became simpler 😄 The files changed of this PR passed from 26 to 15 💯 Thanks!

Refactor done in several commits:
https://github.com/decentraland/unity-explorer/pull/290/commits/3feab5d4b21f0bef35cb302fc0928113e785805d
https://github.com/decentraland/unity-explorer/pull/290/commits/616add76623e3f51606206aa1808d8e8bbd7d348
https://github.com/decentraland/unity-explorer/pull/290/commits/b0be9901fa88769fa93b86e90834fe327f03ea2b

File: Explorer/Assets/DCL/SDKComponents/SceneUI/Classes/DCLTransform.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        public bool IsHidden;
+        public int RightOf;
+
+        private EventCallback<PointerDownEvent> currentOnPointerDownCallback;
+        private EventCallback<PointerUpEvent> currentOnPointerUpCallback;
+
+        public bool HasAnyPointerDownCallback => currentOnPointerDownCallback != null;
+        public bool HasAnyPointerUpCallback => currentOnPointerUpCallback != null;
+
+        public void RegisterPointerDownCallback(EventCallback<PointerDownEvent> newOnPointerDownCallback)
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/SDKComponents/SceneUI/Classes/DCLTransform.cs]: Done: https://github.com/decentraland/unity-explorer/pull/290/commits/d886765b5809e31a0e43c9f1b63470366e33fd49

File: Explorer/Assets/DCL/SDKComponents/SceneUI/Systems/UIPointerEvents/UIPointerEventsSystem.cs (hunk 1 above)
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/SDKComponents/SceneUI/Systems/UIPointerEvents/UIPointerEventsSystem.cs]: As we discussed via slack, I've applied the corresponding changes in order to move the AppendMessage call outside of the callbacks 👍 

Refactor done here: https://github.com/decentraland/unity-explorer/pull/290/commits/5ba26fe5c09e4710962d2b7038a4a9989fea96ab

File: Explorer/Assets/DCL/SDKComponents/SceneUI/Systems/UIPointerEvents/UIPointerEventsSystem.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+using SceneRunner.Scene;
+using UnityEngine.UIElements;
+
+namespace DCL.SDKComponents.SceneUI.Systems.UIPointerEvents
+{
+    [UpdateInGroup(typeof(SyncedSimulationSystemGroup))]
+    [UpdateAfter(typeof(UITransformUpdateSystem))]
+    [LogCategory(ReportCategory.SCENE_UI)]
+    [ThrottlingEnabled]
+    public partial class UIPointerEventsSystem : BaseUnityLoopSystem
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/SDKComponents/SceneUI/Systems/UIPointerEvents/UIPointerEventsSystem.cs:24]: Done: https://github.com/decentraland/unity-explorer/pull/290/commits/ec75d519af32cef73a2b09b31e618f64bc3bb63b

File: Explorer/Assets/DCL/PluginSystem/World/SceneUIPlugin.cs [hunk 5]
```diff
@@ -70,8 +71,9 @@ public void InjectToWorld(ref ArchSystemsWorldBuilder<Arch.Core.World> builder,
             UIInputReleaseSystem.InjectToWorld(ref builder, componentPoolsRegistry);
             UIDropdownInstantiationSystem.InjectToWorld(ref builder, componentPoolsRegistry);
             UIDropdownReleaseSystem.InjectToWorld(ref builder, componentPoolsRegistry);
+            UIPointerEventsSystem.InjectToWorld(ref builder, sharedDependencies.SceneStateProvider, sharedDependencies.EcsToCRDTWriter);
 
-            finalizeWorldSystems.Add(ReleasePoolableComponentSystem<VisualElement, UITransformComponent>.InjectToWorld(ref builder, componentPoolsRegistry));
+            finalizeWorldSystems.Add(ReleasePoolableComponentSystem<UITransformComponent, UITransformComponent>.InjectToWorld(ref builder, componentPoolsRegistry));
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/PluginSystem/World/SceneUIPlugin.cs:76]: Thus this line is not needed either

File: Explorer/Assets/DCL/SDKComponents/SceneUI/Components/UITransformComponent.cs [hunk 6]
```diff
@@ -1,26 +1,52 @@
...
+        {
+            Transform ??= new VisualElement();
+            Transform.name = UiElementUtils.BuildElementName(componentName, entity);
+            Parent = EntityReference.Null;
+            Children = HashSetPool<EntityReference>.Get();
+            IsHidden = false;
+            RightOf = sdkModel.RightOf;
+
+            PointerEventTriggered = null;
+            this.RegisterPointerCallbacks(
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/SDKComponents/SceneUI/Components/UITransformComponent.cs:40]: Subscribing for every transform independently on `PBPointerEvents` is redundant: you will create callbacks for each `UITransform` but it's only needed when `PBPointerEvents` are present.
I propose you to subscribe in a separate function. For example you can store a flag whether subscriptions were done and call  it from `RegisterPointerEvents`

**Review by mikhail-dcl — CHANGES_REQUESTED** [review]
We are almost there. Just some redundancies should be addressed

File: Explorer/Assets/DCL/SDKComponents/SceneUI/Components/UITransformComponent.cs [hunk 7]
```diff
@@ -1,26 +1,52 @@
...
+using Google.Protobuf.Collections;
 using System;
 using System.Collections.Generic;
 using UnityEngine.Pool;
 using UnityEngine.UIElements;
 
 namespace DCL.SDKComponents.SceneUI.Components
 {
-    public struct UITransformComponent : IPoolableComponentProvider<VisualElement>
+    public class UITransformComponent : IPoolableComponentProvider<UITransformComponent>
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/SDKComponents/SceneUI/Components/UITransformComponent.cs]: Oh ok! New knowledge adquired! :) Thanks!

Done here: https://github.com/decentraland/unity-explorer/pull/290/commits/9a2080718bba560e36d06fe903ad7986ee6fa5db

File: Explorer/Assets/DCL/PluginSystem/World/SceneUIPlugin.cs (hunk 5 above)
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/PluginSystem/World/SceneUIPlugin.cs]: Done here: https://github.com/decentraland/unity-explorer/pull/290/commits/9a2080718bba560e36d06fe903ad7986ee6fa5db

File: Explorer/Assets/DCL/SDKComponents/SceneUI/Components/UITransformComponent.cs [hunk 8]
```diff
@@ -1,26 +1,52 @@
...
-        VisualElement IPoolableComponentProvider<VisualElement>.PoolableComponent => Transform;
-        Type IPoolableComponentProvider<VisualElement>.PoolableComponentType => typeof(VisualElement);
+        UITransformComponent IPoolableComponentProvider<UITransformComponent>.PoolableComponent => this;
+        Type IPoolableComponentProvider<UITransformComponent>.PoolableComponentType => typeof(UITransformComponent);
+
+        internal EventCallback<PointerDownEvent> currentOnPointerDownCallback;
+        internal EventCallback<PointerUpEvent> currentOnPointerUpCallback;
+
+        public PointerEventType? PointerEventTriggered;
+        public RepeatedField<PBPointerEvents.Types.Entry> RegisteredPointerEvents { get; internal set; }
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/SDKComponents/SceneUI/Components/UITransformComponent.cs]: You're right, I was storing duplicated data. I've made the corresponding change in order to read that data directly from `PBPointerEvents` 👍 

Done here: https://github.com/decentraland/unity-explorer/pull/290/commits/213d54bb9d6ec094c335ff81c417a3fdb6a547d5

File: Explorer/Assets/DCL/SDKComponents/SceneUI/Components/UITransformComponent.cs (hunk 6 above)
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/SDKComponents/SceneUI/Components/UITransformComponent.cs]: Good point! I've made the subscriptions be made ONLY for UITransforms that also have the `PBPointerEvents` component 👍 Thanks for the suggestion!

Done here: https://github.com/decentraland/unity-explorer/pull/290/commits/975de9f738b3cf0530dbbe14ff5ec811f219d852

**Review by mikhail-dcl — APPROVED** [review]

**Pull request merged**